ligo/src/types.mligo

```diff
@@ -21,6 +21,18 @@ type ledger_key = address * token_id
 type ledger_value = nat
 type ledger = (ledger_key, ledger_value) big_map
 
+// We use this to track the number of tokens that were frozen in each period.
+// When unfreeze is called, we will look into this map to see if there are any
+// frozen tokens that were frozen in a period before the current one. If there
+// are enough frozen tokens from earlier period, then we unfreeze them from the
+// main ledger, and we update the frozen token count for the corresponding
+// previous periods.
+type freeze_history_element = (nat * nat) list
+type freeze_history = (address, freeze_history_element) big_map
+
+// The count of actually staked frozen tokens.
+type staked = (address, nat) big_map
```

Comment:
A couple of things here:

1. we should use a record to know what are the `nat`s
2. I think we don't need a separate `staked` big_map, as only frozen token can be staked

So my suggestion would be to change these types to this:
```ocaml
type freeze_history_entry =
  { period : nat
  ; staked : nat
  ; unstaked : nat
  }

type freeze_history = (address, freeze_history_entry list) big_map
```

and then just use 
```ocaml
  ; freeze_history : freeze_history
```
in the `storage` (removing `staked`).

Aside from being a more correct representation it should also simplify/save some calculations, what do you think?

Reply:
Sounds good.